### PR TITLE
TFP stability diagnostic: map finite recipes in 20-epoch runs

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

TFP's known failure mode is NaN/inf divergence on long runs — not underperformance. Brook's PR #3190 revealed that certain physics features cause gradient explosion at evaluation time (cp_panel wrong index crashes, wake features cause gradient explosion). Before we can optimize metrics on TandemFoil Paper we need to know which recipe components are numerically stable.

This is a **diagnostic mapping exercise**, not a performance optimization. Run 4 cheap 20-epoch probes, each isolating one modification, and report whether each stays finite. The base config is the mandatory TFP configuration (Lion lr=1.25e-4, T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d, Fourier enabled).

## Instructions

**Dataset:** `tandemfoil_paper` only.

**Base config (mandatory for all trials):**
- Optimizer: Lion, lr=1.25e-4
- cosine-t-max=10, grad-clip=0.5, weight-decay=1e-2
- EMA decay=0.999
- 3 layers, 192 hidden dim
- Fourier enabled
- epochs=20 (diagnostic only)
- `--wandb_group tfp-stability-diagnostic`

---

### Trial 1: Baseline stability (no physics features)

Confirm the mandatory base config is stable over 20 epochs with no physics features:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --ema-decay 0.999 \
  --epochs 20 \
  --wandb_group tfp-stability-diagnostic \
  --wandb_run_name tfp-diag-trial1-base
```

Report: (1) stayed finite? (2) val_primary/field_mse at epoch 20 (3) any NaN/inf/gradient explosion?

---

### Trial 2: asinh-pressure + residual-prediction

Add the two transforms that were in the best-known config (#3025):

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --ema-decay 0.999 \
  --asinh-pressure \
  --residual-prediction \
  --epochs 20 \
  --wandb_group tfp-stability-diagnostic \
  --wandb_run_name tfp-diag-trial2-asinh-residual
```

Report: (1) stayed finite? (2) val_primary/field_mse at epoch 20 (3) any NaN/inf/gradient explosion?

---

### Trial 3: asinh-pressure + residual-prediction + LR warmup

Add warmup (3 epochs) to the Trial 2 config to reduce large initial gradient steps:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --ema-decay 0.999 \
  --asinh-pressure \
  --residual-prediction \
  --warmup-epochs 3 \
  --epochs 20 \
  --wandb_group tfp-stability-diagnostic \
  --wandb_run_name tfp-diag-trial3-warmup
```

Report: (1) stayed finite? (2) val_primary/field_mse at epoch 20 (3) any NaN/inf/gradient explosion?

---

### Trial 4: Full best-known config (all physics features from #3025)

Test the complete config from #3025 including cp-panel and pressure-prior-addition:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --ema-decay 0.999 \
  --epochs 20 \
  --wandb_group tfp-stability-diagnostic \
  --wandb_run_name tfp-diag-trial4-full-physics
```

Report: (1) stayed finite? (2) val_primary/field_mse at epoch 20 (3) any NaN/inf/gradient explosion?

---

**Summary table to include in results:**

| Trial | Description | Stayed finite? | val_primary/field_mse@ep20 | NaN/inf/explosion? |
|-------|-------------|---------------|---------------------------|-------------------|
| 1 | Base (no physics) | ? | ? | ? |
| 2 | asinh + residual | ? | ? | ? |
| 3 | asinh + residual + warmup | ? | ? | ? |
| 4 | Full #3025 physics | ? | ? | ? |

## Baseline

- **Dataset:** tandemfoil_paper
- **Primary metric:** `val_primary/field_mse` (lower is better)
- **Current best:** 0.002383 (val) at epoch 443 — PR #3025 (haku)
- **Best W&B run:** d1xh0o1p
- **Reproduce command:**
  ```bash
  cd target/icml2026 && python train.py --dataset tandemfoil_paper --optimizer lion \
      --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier \
      --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame \
      --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction \
      --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
  ```

**Note:** These 20-epoch diagnostic runs will NOT beat 0.002383 — that's expected. The goal is the stability map, not the metric. We need to know which features are safe before extending any run to 999 epochs.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT brook:

## Results

### Summary

The base config (no asinh features) is **perfectly stable** — monotonic convergence over 20 epochs. The `asinh-pressure` transform causes early Infinity but **recovers** within 5 epochs. Adding `--enable-pressure-prior-addition` (with `--enable-cp-panel`) **prevents recovery entirely** due to a wrong-index bug in `cp_panel_prior_index()`. The LR warmup does not help and actually worsens best-case metrics.

### Stability Map

| Trial | Description | Stayed finite? | val_primary/field_mse@ep20 | best val field_mse | NaN/inf/explosion? |
|-------|-------------|---------------|---------------------------|-------------------|--------------------|
| 1 | Base (no physics) | **YES** | 0.0470 | 0.0470 (ep20) | None |
| 2 | asinh + residual | Partially | 4,125,901 | 1,496 (ep5) | Inf ep1 only, then oscillates 10^3–10^10 |
| 3 | asinh + residual + warmup | Partially | 86,933,803 | 19,062 (ep6) | Inf ep2 only, then oscillates 10^4–10^10 |
| 4 | Full #3025 physics | **NO** | Infinity | Infinity (all epochs) | Infinity at every epoch 1-20 |

### Test Metrics (best checkpoint)

| Trial | test_primary/field_mse | Best epoch | W&B run |
|-------|----------------------|------------|---------|
| 1 | **0.0405** | 20 | `2wsrmobl` |
| 2 | 1,262 | 5 | `mj3yhs0n` |
| 3 | 53,238 | 6 | `9dzufksm` |
| 4 | Infinity | 1 | `mdkj5hz8` |

All W&B runs in `wandb-applied-ai-team/senpai-v1`, group `tfp-stability-diagnostic`.

### Per-epoch convergence (Trial 1 — base)

Monotonic improvement every epoch. No instability, no oscillation:
```
ep 1: 0.437  →  ep 5: 0.184  →  ep10: 0.071  →  ep15: 0.055  →  ep20: 0.047
grad_norm: 28-53 (stable)   loss: 1.46 → 0.36 (smooth decline)
```

### Per-epoch convergence (Trial 2 — asinh + residual)

Recovers from Infinity by epoch 3, finds best at epoch 5, then oscillates wildly:
```
ep1: Inf → ep2: 2.7e29 → ep3: 3.3e9 → ep4: 563K → ep5: 1496 (BEST)
ep6-20: oscillates 4,820 – 12,000,000,000 (never improves on ep5 best)
grad_norm: 20-42 (stable)   loss: 0.84 → 0.11 (smooth decline)
```
Key observation: the training loss decreases smoothly, but field_mse oscillates violently. This is because pressure predictions in asinh space are marginally stable — small shifts get amplified exponentially by sinh during eval.

### Per-epoch convergence (Trial 3 — asinh + residual + warmup)

Warmup delays first Infinity to ep2 instead of ep1, but then behaves WORSE than Trial 2:
```
ep1: 4.5M → ep2: Inf → ep3: 4.1e10 → ep5: 1.6e9 → ep6: 19,062 (BEST)
ep7-20: oscillates 38K – 11,000,000,000 (never improves)
```
Best metric is **13x worse** than Trial 2 (19,062 vs 1,496). The warmup's lower initial LR may slow the model's adaptation to asinh space, making it harder to learn the pressure-range constraint.

### Per-epoch convergence (Trial 4 — full #3025 physics)

Infinity at every single epoch despite loss decreasing normally:
```
ep1-20: field_mse = Infinity (all 20 epochs)
loss: 1.02 → 0.097 (smooth decline)
grad_norm: 20-40 (stable)
```
The model IS learning (loss drops 10x, Ux/Uy improve), but the `_apply_pressure_prior_addition` corrupts pressure predictions at every step.

### Root cause: `cp_panel_prior_index()` wrong for tandemfoil_paper

**Why Trial 4 stays at Infinity while Trial 2 recovers:**

The only difference between Trial 2 and Trial 4 is `--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --enable-pressure-prior-addition`. Of these:
- `--enable-te-coord-frame`: no effect on tandemfoil_paper (flag not forwarded to dataset)
- `--enable-cp-panel --enable-cp-panel-tandem-only`: no effect on tandemfoil_paper input tensor (features only added for tandemfoilset)
- `--enable-pressure-prior-addition`: **THIS IS THE PROBLEM**

When both `--enable-cp-panel` and `--enable-pressure-prior-addition` are set, `cp_panel_prior_index()` returns `surface_input_dim - 1`, which points to the **last Fourier feature** (not a cp_panel column, since cp_panel isn't in the input tensor for tandemfoil_paper).

The model then adds this Fourier feature value (~bounded in [-1, 1]) to every pressure prediction in asinh space. Since `sinh(x + δ)` diverges exponentially from `sinh(x)` for large `x`, this constant perturbation prevents the pressure channel from ever producing finite values after sinh inversion.

**Implication for champion reproducibility:** The champion config (PR #3025, run `d1xh0o1p`) uses `--enable-pressure-prior-addition` and achieves field_mse = 0.002383 at epoch 443. The current codebase produces Infinity at all 20 epochs with the same flags. This suggests either (a) the champion was run on a different version of `cp_panel_prior_index()`, or (b) it was run without `--enable-pressure-prior-addition`. The champion's baseline metric may not be reproducible with the current code.

### Bug fix included

Applied the `primary_metric_key` → `_checkpoint_metric_key` rename (same fix as PR #3190) to this branch, since it's also needed here.

### Implemented: `--warmup-epochs` flag

Added `warmup_epochs: int = 0` to `TrainConfig` and wired up `SequentialLR` (LinearLR → CosineAnnealingLR) in the scheduler setup. Warmup linearly ramps LR from ~0 to target over the specified number of epochs.

### Commands used

**Trial 1:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 \
  --model-heads 3 --ema-decay 0.999 --epochs 20 \
  --agent brook --wandb-name "brook/tfp-diag-trial1-base" --wandb-group tfp-stability-diagnostic
```

**Trial 2:**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 \
  --model-heads 3 --ema-decay 0.999 --asinh-pressure --residual-prediction --epochs 20 \
  --agent brook --wandb-name "brook/tfp-diag-trial2-asinh-residual" --wandb-group tfp-stability-diagnostic
```

**Trial 3:**
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 \
  --model-heads 3 --ema-decay 0.999 --asinh-pressure --residual-prediction \
  --warmup-epochs 3 --epochs 20 \
  --agent brook --wandb-name "brook/tfp-diag-trial3-warmup" --wandb-group tfp-stability-diagnostic
```

**Trial 4:**
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 \
  --model-heads 3 --ema-decay 0.999 --asinh-pressure --residual-prediction \
  --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --enable-pressure-prior-addition --epochs 20 \
  --agent brook --wandb-name "brook/tfp-diag-trial4-full-physics" --wandb-group tfp-stability-diagnostic
```

### Suggested follow-ups

1. **Fix `cp_panel_prior_index()` for tandemfoil_paper** — return None when cp_panel isn't in the actual input tensor. This unblocks Trial 4 and restores champion reproducibility.
2. **Run Trial 2 config to 999 epochs** — it recovers from early instability and may converge near the champion level over 443+ epochs. The base config (Trial 1) already reaches 0.047 at epoch 20 without asinh/residual, suggesting the full config could eventually beat it.
3. **Investigate asinh oscillation** — Trials 2 and 3 show smooth loss but wildly oscillating field_mse. The pressure channel is marginally stable in asinh space. A possible fix: use a tighter gradient clip or pressure-specific clamping during training.

**Timestamp:** 2026-04-23 ~12:00 UTC
